Sources/StreamChat/Workers/Background/TokenHandler.swift

```diff
@@ -0,0 +1,225 @@
+//
+// Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+import SwiftUI
+
+/// The type responsible for holding and refreshing the access token.
+protocol TokenHandler: AnyObject {
+    /// The currently used token.
+    var currentToken: Token? { get }
+    
+    /// The user connection provider.
+    var connectionProvider: UserConnectionProvider { get set }
+    
+    /// Assignes the given value to the `currentToken`, propagates it to the waiters, and cancels the ongoing refresh process.
+    /// - Parameter token: The token to set.
+    func set(token: Token, completion: ((Error?) -> Void)?)
+    
+    /// Triggers the token refresh process. When refresh process is completed, updates `currentToken` with the new value and completes token waiters.
+    /// - Parameter completion: The completion that will be called when the token is fetched.
+    func refreshToken(completion: @escaping TokenWaiter)
+    
+    /// Adds the new token waiter without initiating the token refresh process.
+    /// - Parameter completion: The completion that will be called when token is fetched.
+    func add(tokenWaiter: @escaping TokenWaiter) -> WaiterToken
+    
+    /// Removes the waiter with the given token from the list of waiters.
+    /// - Parameter token: The waiter's token
+    func removeTokenWaiter(_ token: WaiterToken)
+    
+    /// Cancels all token waiters with the given error.
+    /// - Parameter error: The error to cancel waiters with.
+    func cancelTokenWaiters(with error: Error)
+}
+
+final class DefaultTokenHandler: TokenHandler {
+    private let maximumTokenRefreshAttempts: Int
+    private var retryStrategy: RetryStrategy
+    private let retryTimerType: Timer.Type
+    private var retryTimer: TimerControl?
+    
+    @Atomic private var isRefreshingToken: Bool = false
+    @Atomic private var tokenWaiters: [WaiterToken: TokenWaiter] = [:]
```

Comment:
Might be better to handle it using dispatch queue, left as it was for simplicity